internal/services/compute/virtual_machine_implicit_data_disk_from_source.go

```diff
@@ -0,0 +1,476 @@
+					if profile := props.StorageProfile; profile != nil {
+						if dataDisks := profile.DataDisks; dataDisks != nil {
+							for _, dataDisk := range *dataDisks {
+								if pointer.From(dataDisk.Name) == id.Name {
+									disk = &dataDisk
+									break
+								}
+							}
+						}
+					}
+				}
+			}
+
+			if disk == nil {
+				return metadata.MarkAsGone(*id)
+			}
+
+			schema.Lun = disk.Lun
+			if v := pointer.From(disk.Caching); v != virtualmachines.CachingTypesNone {
+				schema.Caching = string(v)
+			}
+
+			schema.CreateOption = string(disk.CreateOption)
+			schema.DeleteOption = string(pointer.From(disk.DeleteOption))
+			schema.DiskSizeGb = pointer.From(disk.DiskSizeGB)
+			if disk.SourceResource != nil {
+				schema.SourceResourceId = pointer.From(disk.SourceResource.Id)
+			}
+
+			schema.WriteAcceleratorEnabled = pointer.From(disk.WriteAcceleratorEnabled)
+
+			return metadata.Encode(&schema)
+		},
+	}
+}
+
+func (r VirtualMachineImplicitDataDiskFromSourceResource) Delete() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.Compute.VirtualMachinesClient
+
+			id, err := parse.DataDiskID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			locks.ByName(id.VirtualMachineName, VirtualMachineResourceName)
+			defer locks.UnlockByName(id.VirtualMachineName, VirtualMachineResourceName)
+
+			virtualMachineId := virtualmachines.NewVirtualMachineID(id.SubscriptionId, id.ResourceGroup, id.VirtualMachineName)
+			resp, err := client.Get(ctx, virtualMachineId, virtualmachines.DefaultGetOperationOptions())
+			if err != nil {
+				return fmt.Errorf("retrieving %s: %+v", virtualMachineId, err)
+			}
+
+			if model := resp.Model; model != nil {
+				if props := model.Properties; props != nil {
+					if profile := props.StorageProfile; profile != nil {
+						newDisks := make([]virtualmachines.DataDisk, 0)
+						var toBeDeletedDisk *virtualmachines.DataDisk
+						if dataDisks := profile.DataDisks; dataDisks != nil {
+							for _, dataDisk := range *dataDisks {
+								if pointer.From(dataDisk.Name) != id.Name {
+									newDisks = append(newDisks, dataDisk)
+								} else {
+									toBeDeletedDisk = pointer.To(dataDisk)
+								}
+							}
+						}
+
+						profile.DataDisks = &newDisks
+
+						// fixes #24145
+						model.Properties.ApplicationProfile = nil
+
+						// fixes #2485
+						model.Identity = nil
+						// fixes #1600
+						model.Resources = nil
+
+						err = client.CreateOrUpdateThenPoll(ctx, virtualMachineId, *model, virtualmachines.DefaultCreateOrUpdateOperationOptions())
+						if err != nil {
+							return fmt.Errorf("deleting %s: %+v", id, err)
+						}
+
+						// delete data disk if delete_option is set to delete
+						if toBeDeletedDisk != nil && pointer.From(toBeDeletedDisk.DeleteOption) == virtualmachines.DiskDeleteOptionTypesDelete &&
```

Comment:
Because Terraform is creating this resource, I believe we can make the assumption that Terraform can delete this resource? If there is value in keeping the disk around, we can add a new variable to the  `features` block that will hang on to the disk instead of deleting it. Or vice versa if the default should be to keep the disk around instead

Reply:
This resource is a little special. It manages the [data disk](https://learn.microsoft.com/en-us/rest/api/compute/virtual-machines/create-or-update?view=rest-compute-2024-03-01&tabs=HTTP#datadisk) of a VM such as `azurerm_virtual_machine_data_disk_attachment`. When creating this resource, what is really done is to add a data disk configuration to the existing VM, which will trigger the Azure service to create a new data disk for the VM. When deleting this resource, what is really done is to remove the data disk configuration from the existing VM. However, the physical data disk will not be removed from the resource group, which may cause the resource group cannot be deleted later. That's why I add this ugly delete option. According to [your comment](https://github.com/hashicorp/terraform-provider-azurerm/pull/25537#discussion_r1607415824),  do you mean we can hang on to the disk by default and should add a new variable to the provider [features block](https://github.com/hashicorp/terraform-provider-azurerm/blob/343bf3650e0da580c30f8b15d843f7288c4c8314/internal/provider/provider.go#L320) to control whether to delete the physical data disk when deleting this resource?

Reply:
This is tough because from the users perspective, Terraform is creating a disk (I do see that Azure is creating the disk from what we are sending to the api but the user will see that it is Terraform that is creating the disk) so running destroy should delete the disk that was created. 

My thought is to have the default be to delete this disk and then in a features block, we add an option to detach the disk instead of deleting it.